app/models/access_grant.rb

```diff
@@ -22,30 +22,75 @@ def self.authenticate(code, client_id)
     AccessGrant.where("code = ? AND client_id = ?", code, client_id).first
   end
 
-  # Pretty much perform the 1st step of the OAuth2 authorization.
-  def self.get_authorize_redirect_uri(user, params)
-    client = Client.find_by_app_id(params[:client_id])
-    unless client
-      raise "Client not found"
+  ValidationResult = Struct.new(:valid, :client, :error_redirect) do
+    def valid?
+      valid
+    end
+
+    def error(error_msg, redirect_uri)
+      self.error_redirect = client.get_redirect_uri(redirect_uri, error: error_msg)
+      self.valid = false
     end
-    unless SUPPORTED_RESPONSE_TYPES.include?(params[:response_type])
+  end
+
+  def self.matching_response_type(client_type, response_type)
+    # Implicit flow for public clients (e.g. Glossary Authoring).
+    (client_type == Client::PUBLIC && response_type === "token") ||
+    # Auth code flow (two steps) for confidential clients (e.g. LARA).
+    (client_type == Client::CONFIDENTIAL && response_type === "code")
+  end
+
+  # this will raise an error if:
+  # - the client is not found
+  # - the passed in redirect_uri is not registered with the client
```

Comment:
I could add a comment that validation might fail in two different ways:
- "hard", when redirect_uri is malformed or not registered. We raise an error then and user sees 500 error.
- "soft", when redirect_uri is fine. The user will be redirected back with the error message provided to the client.

I guess if someone isn't super familiar with oauth, he might get a bit surprised why we have these two paths.

Reply:
Good point. I updated the comment.